### DISCUSSION TITLE
How to parse this minimalist program?

### DISCUSSION THREAD

**Discussion by ethindp** [taocpp/PEGTL#227] [Q&A]
I'm learning PEGs by trying out a basic pascal-ish language. I've got the following grammar:

```cpp
struct Word: pad<sor<plus<identifier>>, space> { };
struct BasicReal: rep_min_max<1, 1, star<digit>, string<'.'>,plus<digit>> { };
struct Exponent: sor<string<'e'>, string<'E'>> { };
struct OptNegPos: opt<sor<string<'+'>, string<'-'>>> {};
struct RealDigits: rep_min_max<3, 3, digit> { };
struct ScientificForm: rep_min_max<1, 1, sor<rep_min_max<1, 1, Exponent, OptNegPos, RealDigits>>> { };
struct Real: pad<sor<BasicReal, rep_min_max<1, 1, BasicReal, ScientificForm>>, space> { };
struct Special: pad<sor<string<','>, string<'('>, string<')'>, string<'['>, string<']'>, string<';'>, string<'<'>, string<'>'>, string<'*'>, string<'/'>, string<'+'>, string<'-'>, string<'='>>, space> { };
struct Integer: pad<plus<digit>, space>{ };
struct HexInteger: pad<rep_min_max<1, 1, string<'0', 'x'>, plus<xdigit>>, space> { };
struct Grammar: must<opt<sor<bof, bol>>, until<plus<sor<Word, Real, Special, Integer, HexInteger>>>> { };
```
All my code does right now is tokenize the input, but I keep getting a parse error when I try to evaluate the following code:
```
program abc
integer a,b,c
real x
begin
    b=6;
    a=b+-4;
    x=-5.63e002;
end
end
```
It always breaks on the final line, either the "end" or, if I add an extra newline, on that newline. Is there something I'm missing?

**Comment by d-frey** [comment]
Your grammar contains a lot of weird things, e.g. `sor<plus<identifier>>` (part of the first rule) matches `identifier`. Both the `sor` and the `plus` are useless. Since it seems you are starting with writing grammars, I suggest you take a look at our [tracer](https://github.com/taocpp/PEGTL/blob/master/doc/Getting-Started.md#tracer), which can help you to verify whether and how your grammar matches to your input.

**Comment by ColinH** [comment]
How are things going, have you managed to make some progress? Any further questions or issues that you might want to discuss?

**Comment by ethindp** [comment]
Yes, I have. I don't have any further questions at this time. I
probably will sometime later though. :)

On 11/19/20, Dr. Colin Hirsch <notifications@github.com> wrote:
> How are things going, have you managed to make some progress? Any further
> questions or issues that you might want to discuss?
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/taocpp/PEGTL/issues/227#issuecomment-730576767


-- 
Signed,
Ethin D. Probst
